rfcs/12-primitive-shapes.md

```diff
@@ -0,0 +1,514 @@
+
+ Translation and rotation are **not** defined using bevy's `Transform` components. This is because types that implement `Bounding` and `Collider` must be fully self-contained. This:
+ 
+ * makes the API simpler when using these components in functions and systems
+ * ensures bounding and collision types use an absolute minimum of memory
+ * prevents errors caused by nonuniform scale invalidating the shape of the primitive.
+
+Instead, changes to the parent's `GlobalTransform` should be used to derive a new primitive on scale changes - as well as a new translation and rotation if required.
+
+### Meshing
+
+Both 2d and 3d primitives can implement the `Meshable` trait to provide the ability to generate a tri mesh:
+
+```rust
+let circle_mesh: Mesh = Circle2d{ radius: 2.0 }.mesh();
+```
+The base primitive types only define the shape (`Circle2d`) and size (`radius`) of the geometry about the origin. Once generated, the mesh can have a transform applied to it like any other mesh, and it is no longer tied to the primitive that generated it. The `Default::default()` implementation of primitives should be a "unit" variant, such as a circle with diameter 1.0, or a box with all edges of length 1.0.
+
+Meshing could be naturally extended with other libraries or parameters. For example, a sphere by default might use `Icosphere`, but could be extended to generate UV spheres or quad spheres. For 2d types, we can use a crate such as `lyon` to generate 2d meshes from parameterized primitive shapes within the `Meshable` interface.
+
+### Bounding
+
+For use in spatial queries, broad phase collision detection, and raycasting, `Bounding` and `Bounding2d` traits are provided. These traits are implemented for types that define position in addition to the shape and size defined in the base primitive. The basic functionality of this trait is to check whether one bounding shape is contained within another bounding shape.
+
+### Colliders
+
+Colliders can provide intersection checks against other colliders, as well as check if they are within a bounding volume.
+
+### 3D and 2D
+
+This RFC provides independent 2d and 3d primitives. Recall that the purpose of this is to provide lightweight types, so there are what appear to be duplicates in 2d and 3d, such as `Line` and `Line2d`. Note that the 2d version of a line is smaller than its 3d counterpart because it is only defined in 2d. 3d geometry (or 2d with depth) is assumed to be the default for most cases. The names of the types were chosen with this in mind.
+
+## Implementation strategy
+
+### Helper Types
+
+```rust
+/// Stores an angle in radians, and supplies builder functions to prevent errors (from_radians, from_degrees)
+struct Angle(f32);
+```
+### Traits
+
+These traits are provided as reference to illustrate how these might be used. The details of implementation and interface should be determined in a separate RFC, PR, or independent prototypes.
+
+```rust
+trait Meshable{
+  fn mesh(&self) -> Mesh;
```

Comment:
I think it's likely users will need to control how the mesh is generated, e.g. how many segments in a circle. Using a trait across all shapes might make that difficult?

Reply:
Yes. I addressed this elsewhere in the doc, but the point of this trait was to provide a simple, sane, default mesh, especially for immediate mode debug uses. I considered adding a `subdivisions: u8` field, but decided to keep it simple. I also wanted to avoid specifying too much detail in meshing, as that is a whole other can of worms and I wanted to keep that out of the scope of this RFC. At the very least, this would provide the base primitives we can extend with shape-specialized meshing functions, like icosphere/uvsphere/box for the `Sphere` type.

Reply:
I would include this trait but implement it using a `MeshBuilder` struct with static functions elsewhere, e.g.

```rust
impl Meshable for Sphere {
    fn mesh(&self) -> Mesh {
        let num_segments = 12; // Default args for sphere mesh
        MeshBuilder::build_sphere(&self, num_segments, /* other sphere specific arguments */);
    }
}
```

This way the underlying mesh creation functions are always available to the user in the exact same manner as `mesh` -- call `MeshBuilder` directly with the relevant shape function (`MeshBuilder::build_cube`) and provide whatever shape specific arguments are needed.

Reply:
Right, the shape primitives are entirely distinct from meshes that can be used to represent them.

Based on your other feedback, I think we are heading toward:

1. `impl From<SomeShape> for Mesh` generates a sane default mesh for debugging and prototyping purposes.
2. When you _do_ care about the mesh's construction, we could provide something like `impl From<Sphere> for Icosphere` and `UvSphere`. In my mind this cements the idea that a primitive `Sphere` is _entirely distinct_ from a `UvSphere` or a `UvSphere`. The former is a precise geometric primitive, the latter are types of meshes with their own special parameterization.

Reply:
I don't think (1) is compatible with what I said above (which is an argument against providing "default" implementations for things that aren't meshes). I'm arguing that there _is_ no sane default mesh for a sphere. Every context is different.

I think "debugging" something like a Circle should involve higher level orchestration because there _is_ no universal debugging context:

```rust
// convert the circle to a polygonal mesh with fill faces.
// maybe the debug context is drawing a ton of circles and vertex_count needs to be low to cut down on total geometry counts
// maybe the debug context needs very accurate circles to help debug things like collisions, so vertex_count needs to be very high.
Mesh::from(circle.get_polygon(vertex_count));

// convert the circle to a polygonal mesh without fill faces
Mesh::from(circle.get_unfilled_polygon(vertex_count));

// draw the circle using some immediate mode pixel-perfect api
im_shapes::draw_perfect_circle(circle)
```

Reply:
If my understanding is correct. We'd be getting several categories of shape:
* "Abstract" shapes that are simple geometrical descriptions. The same shape could have multiple abstract representations. Each abstract shape may have an infinity of concrete representation, a single one or none at all.
* "Concrete" shapes, with a 1-to-1 mesh correspondence. Though, you could have "circle" vs "disc" which quite obviously describe two different things and will have different meshes.

The current `bevy::shapes` library seems to go for the "Concrete" kind of shapes, with a description of the number of latitudes and longitudes in the `Capsule` struct for example.

The shapes library could support any arbitrary translation from Abstract to Abstract or Abstract to Concrete (with a density parameter, surely) implemented, not as `Into<Foo>` trait, but as simple methods on each shape structs. There would be a rich variety of traits such as each `Concrete` types could be converted into something concrete that can be used such as:
* `Meshable { fn mesh(self) -> Mesh }`
* `Outlinable { fn outline(self) -> Outline }`
* `Collidable { // etc.`

I think such an API would be exhaustive as cart wants, but I also think it's going to be unwieldy knowing what converts to what, you'll have to hunt the conversion methods or trait implementation in the doc on the page of each struct, which is not optimal, but status-quo currently for rust doc community.

For the end user, now this would look like the following:
```rust
// where 34 is the number of subdivisions
Mesh::from(Sphere { radius: 3.0 }.isosphere(34))
```
It's not as terse as the proposed solution. And I'm worried about the discoverability of the conversion methods, but I think it's still quite clean.